[aot_inductor] Lightweight model runner

bertmaher opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #110158

It's useful to have a simple, lightweight way to run a model that adds
essentially no overhead to calling the model's generated `run_impl` method.
This C API is a super thin wrapper around AOTInductorModel: Create, Run, and
Delete are provided, and do very little work beyond dispatch to the appropriate
helpers.

Note the Create function also provides additional functionality beyond the
Container API; it allows the user to pass in a weight map defined in userland,
which is a requirement for several serving use cases.

Differential Revision: [D49670711](https://our.internmc.facebook.com/intern/diff/D49670711/)

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler

bertmaher committed:
[aot_inductor] Lightweight model runner

It's useful to have a simple, lightweight way to run a model that adds
essentially no overhead to calling the model's generated `run_impl` method.
This C API is a super thin wrapper around AOTInductorModel: Create, Run, and
Delete are provided, and do very little work beyond dispatch to the appropriate
helpers.

Note the Create function also provides additional functionality beyond the
Container API; it allows the user to pass in a weight map defined in userland,
which is a requirement for several serving use cases.

Differential Revision: [D49670711](https://our.internmc.facebook.com/intern/diff/D49670711/)

[ghstack-poisoned]

desertfire commented on torch/csrc/inductor/aoti_runtime/interface.h:
```diff
@@ -112,4 +118,15 @@ AOTIRuntimeError AOTInductorModelContainerGetMaxOutputShape(
     const int64_t** ret_output_sizes,
     int64_t* ret_output_ndim);
 
+AOTIRuntimeError AOTInductorModelCreate(
```
I would prefer to have the naming more explicitly state how it is different from AOTInductorModelContainerCreate, or at least add comments here to explain that.

bertmaher replied:
Can you suggest a name?  IMO this is pretty self explanatory so I'm struggling to come up with something satisfactory.

chenyang78 replied:
Perhaps we could just add comments to explain when the user would prefer to use this function, e.g. like those sentences from the summary?

desertfire reviewed (approved):
Adding comments is reasonable to me.

bertmaher committed:
Add comments describing the use case vs AOTInductorModelContainer on "[aot_inductor] Lightweight model runner"


It's useful to have a simple, lightweight way to run a model that adds
essentially no overhead to calling the model's generated `run_impl` method.
This C API is a super thin wrapper around AOTInductorModel: Create, Run, and
Delete are provided, and do very little work beyond dispatch to the appropriate
helpers.

Note the Create function also provides additional functionality beyond the
Container API; it allows the user to pass in a weight map defined in userland,
which is a requirement for several serving use cases.

Differential Revision: [D49670711](https://our.internmc.facebook.com/intern/diff/D49670711/)

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 aakhundov ColinPeppler

[ghstack-poisoned]
